cpmpy/solvers/cplex.py

```diff
@@ -0,0 +1,459 @@
+    $ pip install docplex
+
+    docplex documentation:
+    https://ibmdecisionoptimization.github.io/docplex-doc/
+    You will also need to install CPLEX Optimization Studio from IBM's website.
+    There is a free community version available.
+    https://www.ibm.com/products/ilog-cplex-optimization-studio
+    See detailed installation instructions at:
+    https://www.ibm.com/docs/en/icos/22.1.2?topic=2212-installing-cplex-optimization-studio
+    Creates the following attributes (see parent constructor for more):
+    - cplex_model: object, CPLEX model object
+    """
+
+    _domp = None  # Static attribute to hold the docplex.cp module
+
+    @classmethod
+    def get_domp(cls):
+        if cls._domp is None:
+            import docplex.mp as domp  # Import only once
+            cls._domp = domp
+        return cls._domp
+
+    @staticmethod
+    def supported():
+        return CPM_cplex.installed() and CPM_cplex.license_ok()
+
+    @staticmethod
+    def installed():
+        # try to import the package
+        try:
+            import docplex.mp as domp
+            return True
+        except ModuleNotFoundError:
+            return False
+
+    @staticmethod
+    def license_ok():
+        if not CPM_cplex.installed():
+            warnings.warn(
+                f"License check failed, python package 'docplex' is not installed! Please check 'CPM_cplex.installed()' before attempting to check license.")
+            return False
+        else:
+            try:
+                from docplex.mp.model import Model
+                mdl = Model()
+                mdl.solve()
+                return True
+            except Exception as e:
+                warnings.warn(f"Problem encountered with CPLEX installation: {e}")
+
+    def __init__(self, cpm_model=None, subsolver=None):
+        """
+        Constructor of the native solver object
+
+        Arguments:
+        - cpm_model: a CPMpy Model()
+        - subsolver: None, not used
+        """
+        if not self.installed():
+            raise Exception("CPM_cplex: Install the python package 'docplex' to use this solver interface.")
+        elif not self.license_ok():
+            raise Exception("CPM_cplex: A problem occured during license check. Make sure your installed the CPLEX Optimization Studio")
+        import docplex.mp.model as dmm
+
+        self.cplex_model = dmm.Model()
+        super().__init__(name="cplex", cpm_model=cpm_model)
+
+    @property
+    def native_model(self):
+        """
+            Returns the solver's underlying native model (for direct solver access).
+        """
+        return self.cplex_model
+
+
+    def solve(self, time_limit=None, **kwargs):
+        """
+            Call the cplex solver
+
+            Arguments:
+            - time_limit:  maximum solve time in seconds (float, optional)
+            - kwargs:      any keyword argument, sets parameters of solver object
+
+            Examples of supported arguments include:
```

Comment:
These long-lines should be shortened massively... Just a concise description of what the parameter does is enough I feel. E.g., the `log_output` can be just the `solver logs are written to stdout` or smt. If a user wants to know the details of a parameter there is a nice link ; )

Reply:
We often also add the parameter that controls the number of threads the solver uses. I think such a parameter is more releavant to a use than e.g., `context`